rubix-core/src/main/java/com/qubole/rubix/core/NonLocalReadRequestChain.java

```diff
@@ -12,24 +12,143 @@
+        this.conf = conf;
+        this.clusterType = clusterType;
     }
 
     public ReadRequestChainStats getStats()
     {
         return new ReadRequestChainStats()
-                .setRemoteReads(requests)
                 .setNonLocalReads(requests)
-                .setRequestedRead(totalRead)
+                .setRequestedRead(totalRead + directRead)
                 .setNonLocalDataRead(totalRead);
     }
+
+    @Override
+    public Integer call()
+            throws Exception
+    {
+        Thread.currentThread().setName(threadName);
+        if (readRequests.size() == 0) {
+            return 0;
+        }
+        checkState(isLocked, "Trying to execute Chain without locking");
+        SocketChannel transferClient;
+
+        try {
+            BookKeeperFactory bookKeeperFactory = new BookKeeperFactory();
+            transferClient = bookKeeperFactory.createTransferClient(remoteNodeName, conf);
+        }
+        catch (Exception e) {
+            log.info("Could not create Transfer Client " + Throwables.getStackTraceAsString(e));
+            return directReadRequest(0);
+        }
+
+        for (ReadRequest readRequest : readRequests) {
+            int nread = 0;
+            ByteBuffer buf = ByteBuffer.allocate(CacheConfig.getDataTransferBufferSize(conf));
+            /* order is: long : offset, int : readLength, long : fileSize, long : lastModified,
+                int : clusterType, int : filePathLength, String : filePath */
+            buf.putLong(readRequest.getActualReadStart());
+            buf.putInt(readRequest.getActualReadLength());
+            buf.putLong(fileSize);
+            buf.putLong(lastModified);
+            buf.putInt(clusterType);
+            buf.putInt(filePath.length());
+            buf.put(filePath.getBytes());
+            buf.flip();
+            if (transferClient.isConnected()) {
+                try {
+                    transferClient.write(buf);
+                }
+                catch (IOException e) {
+                    e.printStackTrace();
+                    transferClient.close();
+                    return directReadRequest(readRequests.indexOf(readRequest));
+                }
+            }
+
+            int bytesread = 0;
+            ByteBuffer dst = ByteBuffer.wrap(readRequest.destBuffer, readRequest.getDestBufferOffset(), readRequest.destBuffer.length - readRequest.getDestBufferOffset());
+
+            while (bytesread != readRequest.getActualReadLength()) {
+                if (transferClient.isConnected()) {
+                    try {
+                        nread = transferClient.read(dst);
+                        bytesread += nread;
+                        totalRead += nread;
+                    }
+                    catch (IOException e) {
+                        e.printStackTrace();
+                        return directReadRequest(readRequests.indexOf(readRequest));
+                    }
+                }
+                if (nread == -1) {
```

Comment:
nread is the dataRead for the current loop. It's -1 when connection is closed abruptly.